File: multiVendorEProject/src/main/java/com/abc/multiVendorEProject/service/DealService.java
package com.abc.multiVendorEProject.service;

import com.abc.multiVendorEProject.DTOs.projectDtos.DealRequestDto;
import com.abc.multiVendorEProject.DTOs.projectDtos.DealResponseDto;
import com.abc.multiVendorEProject.entity.Deal;
import com.abc.multiVendorEProject.entity.Product;
import com.abc.multiVendorEProject.mapper.DealMapper;
import com.abc.multiVendorEProject.repository.DealRepository;
import com.abc.multiVendorEProject.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional
public class DealService {

    private final DealRepository dealRepository;
    private final ProductRepository productRepository;
    private final DealMapper dealMapper;

    // ==========================================
    // Get All Deals (Admin)
    // ==========================================
    @Transactional(readOnly = true)
    public Page<DealResponseDto> getAllDeals(
            int page,
            int size,
            String sortBy,
            String sortDir) {

        return dealRepository
                .findAll(createPageable(page, size, sortBy, sortDir))
                .map(dealMapper::toDto);
    }

    // ==========================================
    // Get Deal By Id
    // ==========================================
    @Transactional(readOnly = true)
    public DealResponseDto getDealById(Long id) {

        Deal deal = dealRepository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Deal not found with id: " + id));

        return dealMapper.toDto(deal);
    }

    // ==========================================
    // Create Deal
    // ==========================================
    public DealResponseDto createDeal(DealRequestDto dto) {

        Product product = productRepository.findById(dto.productId())
                .orElseThrow(() ->
                        new RuntimeException("Product not found with id: " + dto.productId()));

        Deal deal = dealMapper.toEntity(dto);

        deal.setProduct(product);

        Deal savedDeal = dealRepository.save(deal);

        return dealMapper.toDto(savedDeal);
    }

    // ==========================================
    // Update Deal
    // ==========================================
    public DealResponseDto updateDeal(Long id, DealRequestDto dto) {

        Deal deal = dealRepository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Deal not found with id: " + id));

        Product product = productRepository.findById(dto.productId())
                .orElseThrow(() ->
                        new RuntimeException("Product not found with id: " + dto.productId()));

        dealMapper.updateEntity(deal, dto);

        deal.setProduct(product);

        Deal updatedDeal = dealRepository.save(deal);

        return dealMapper.toDto(updatedDeal);
    }

    // ==========================================
    // Delete Deal
    // ==========================================
    public void deleteDeal(Long id) {

        Deal deal = dealRepository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Deal not found with id: " + id));

        dealRepository.delete(deal);
    }

    // ==========================================
    // Active Deals (Home Page)
    // ==========================================
    @Transactional(readOnly = true)
    public Page<DealResponseDto> getActiveDeals(
            int page,
            int size,
            String sortBy,
            String sortDir) {

        return dealRepository
                .findActiveDeals(createPageable(page, size, sortBy, sortDir))
                .map(dealMapper::toDto);
    }

    // ==========================================
    // Pageable Helper
    // ==========================================
    private Pageable createPageable(
            int page,
            int size,
            String sortBy,
            String sortDir) {

        Sort sort = sortDir.equalsIgnoreCase("desc")
                ? Sort.by(sortBy).descending()
                : Sort.by(sortBy).ascending();

        return PageRequest.of(page, size, sort);
    }


    @Transactional(readOnly = true)
    public DealResponseDto getActiveDealByProduct(Long productId) {

        Deal deal = dealRepository.findActiveDealByProductId(productId)
                .orElseThrow(() ->
                        new RuntimeException("No active deal found for this product"));

        return dealMapper.toDto(deal);
    }

}

File: e-AngularSide/src/app/services/deal.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from './environments';
import {
  DealRequest,
  DealResponse,
  DealPageResponse
} from '../models/deal.model';

@Injectable({
  providedIn: 'root'
})
export class DealService {

  private readonly baseUrl =
    `${environment.apiUrl}/admin/deals`;

  constructor(
    private http: HttpClient
  ) { }

  // =========================================
  // Admin - Get All Deals
  // =========================================
  getAllDeals(
    page = 0,
    size = 10,
    sortBy = 'createdAt',
    sortDir = 'desc'
  ): Observable<DealPageResponse> {

    const params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sortBy', sortBy)
      .set('sortDir', sortDir);

    return this.http.get<DealPageResponse>(
      this.baseUrl,
      { params }
    );

  }

  // =========================================
  // Get Deal By Id
  // =========================================
  getDealById(id: number): Observable<DealResponse> {

    return this.http.get<DealResponse>(
      `${this.baseUrl}/${id}`
    );

  }

  // =========================================
  // Create Deal
  // =========================================
  createDeal(
    request: DealRequest
  ): Observable<DealResponse> {

    return this.http.post<DealResponse>(
      this.baseUrl,
      request
    );

  }

  // =========================================
  // Update Deal
  // =========================================
  updateDeal(
    id: number,
    request: DealRequest
  ): Observable<DealResponse> {

    return this.http.put<DealResponse>(
      `${this.baseUrl}/${id}`,
      request
    );

  }

  // =========================================
  // Delete Deal
  // =========================================
  deleteDeal(id: number): Observable<string> {

    return this.http.delete(
      `${this.baseUrl}/${id}`,
      {
        responseType: 'text'
      }
    );

  }

  // =========================================
  // Public - Active Deals
  // =========================================
  getActiveDeals(
    page = 0,
    size = 10,
    sortBy = 'createdAt',
    sortDir = 'desc'
  ): Observable<DealPageResponse> {

    const params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sortBy', sortBy)
      .set('sortDir', sortDir);

    return this.http.get<DealPageResponse>(
      `${this.baseUrl}/active`,
      { params }
    );

  }

  // =========================================
  // Public - Active Deal By Product
  // =========================================
  getActiveDealByProduct(
    productId: number
  ): Observable<DealResponse> {

    return this.http.get<DealResponse>(
      `${this.baseUrl}/product/${productId}`
    );

  }

}